packages/nft-provider/lib/NftProvider.ts

```diff
@@ -0,0 +1,112 @@
+import { NodeProvider } from '@liquality/node-provider'
+import { addressToString } from '@liquality/utils'
+import { ensure0x } from '@liquality/ethereum-utils'
+import { NftProvider as INftProvider, Address } from '@liquality/types'
+
+import { NftBaseProvider } from '@liquality/nft-base-provider'
+import { NftErc721Provider } from '@liquality/nft-erc721-provider'
+import { NftErc1155Provider } from '@liquality/nft-erc1155-provider'
+
+type AssetMap = Record<string, Partial<INftProvider> & NftBaseProvider>
+
+export default class NftProvider extends NodeProvider implements INftProvider {
+  _subProviders: AssetMap
+  _nftContractsCache: AssetMap
+  _clientAttached: boolean
+
+  constructor(apiURI: string, apiKey = '') {
+    super({
+      baseURL: apiURI,
+      responseType: 'text',
+      transformResponse: undefined,
+      headers: {
+        'X-Api-Key': apiKey
+      }
+    })
+
+    this._subProviders = {}
+    this._subProviders['ERC721'] = new NftErc721Provider()
+    this._subProviders['ERC1155'] = new NftErc1155Provider()
+
+    this._nftContractsCache = {}
+    this._clientAttached = false
+  }
+
+  async balance(contract: Address | string, tokenIDs: number | number[]) {
+    this._setClientToSubProviders()
```

Comment:
Why does this happen before each function instread of just once?
I think you can override `setProvider` to do this very simply